src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemorySegmentImpl.java

```diff
@@ -285,4 +308,88 @@ private int id() {
         return Math.abs(Objects.hash(base, min, NONCE));
     }
 
+    static class SegmentSplitter implements Spliterator<MemorySegment> {
+        MemorySegmentImpl segment;
+        long elemCount;
+        final long elementSize;
+        long currentIndex;
+
+        SegmentSplitter(long elementSize, long elemCount, MemorySegmentImpl segment) {
+            this.segment = segment;
+            this.elementSize = elementSize;
+            this.elemCount = elemCount;
+        }
+
+        @Override
+        public SegmentSplitter trySplit() {
+            if (currentIndex == 0 && elemCount > 1) {
+                MemorySegmentImpl old = segment;
+                long rem = elemCount % 2;
+                elemCount  = elemCount / 2;
+                long lobound = elemCount * elementSize;
+                long hibound = lobound + (rem * elementSize);
+                segment = old.asSliceNoCheck(0, lobound);
+                return new SegmentSplitter(elementSize, elemCount + rem, old.asSliceNoCheck(lobound, hibound));
+            } else {
+                return null;
+            }
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super MemorySegment> action) {
+            Objects.requireNonNull(action);
+            if (currentIndex < elemCount) {
+                MemorySegmentImpl acquired = segment.acquire();
+                try {
+                    action.accept(acquired.asSliceNoCheck(currentIndex * elementSize, elementSize));
+                } finally {
+                    acquired.closeNoCheck();
+                }
+                currentIndex++;
+                if (currentIndex == elemCount) {
+                    segment = null;
+                }
+                return true;
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public void forEachRemaining(Consumer<? super MemorySegment> action) {
+            Objects.requireNonNull(action);
+            if (currentIndex < elemCount) {
+                MemorySegmentImpl acquired = segment.acquire();
```

Comment:
I would ensure the spliterator state is consumed before performing the bulk loop (take local copies before updating and then looping).  Thereby if the consumer throws an exception it does not leave the spliterator in a partial state i.e. it is guaranteed to be fully traversed if subsequently operated on.  In the current implementation i believe an NPE will be thrown.